group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,903 @@
+     * <li> If there was an assignment that wasn't retained due to a rack mismatch,
+     *      check if the partition can retain its existing assignment. </li>
+     * <li> Sort members in ascending order of their current target assignment sizes
+     *      to ensure the least filled member gets the partition first. </li>
+     */
+    private void unassignedPartitionsAssignment() {
+        List<TopicIdPartition> sortedPartitions = sortTopicIdPartitions(unassignedPartitions);
+
+        for (TopicIdPartition partition : sortedPartitions) {
+            if (rackInfo.useRackStrategy && currentPartitionOwners.containsKey(partition)) {
+                String prevOwner = currentPartitionOwners.get(partition);
+                if  (assignmentManager.maybeAssignPartitionToMember(partition, prevOwner)) {
+                    continue;
+                }
+            }
+
+            TreeSet<String> sortedMembers = assignmentManager.sortMembersByAssignmentSize(
+                membersPerTopic.get(partition.topicId())
+            );
+
+            for (String member : sortedMembers) {
+                if (assignmentManager.maybeAssignPartitionToMember(partition, member)) {
+                    break;
+                }
+            }
+        }
+    }
+
+    /**
+     * If a topic has two or more potential members it is subject to reassignment.
+     *
+     * @return true if the topic can participate in reassignment, false otherwise.
+     */
+    private boolean canTopicParticipateInReassignment(Uuid topicId) {
+        return membersPerTopic.get(topicId).size() >= 2;
+    }
+
+    /**
+     * If a member is not assigned all its potential partitions it is subject to reassignment.
+     * If any of the partitions assigned to a member is subject to reassignment, the member itself
+     * is subject to reassignment.
+     *
+     * @return true if the member can participate in reassignment, false otherwise.
+     */
+    private boolean canMemberParticipateInReassignment(String memberId) {
+        Set<Uuid> assignedTopicIds = targetAssignment.get(memberId).targetPartitions().keySet();
+
+        int currentAssignmentSize = assignmentManager.targetAssignmentSize(memberId);
+        int maxAssignmentSize = assignmentManager.maxAssignmentSize(memberId);
+
+        if (currentAssignmentSize > maxAssignmentSize)
+            LOG.error("The member {} is assigned more partitions than the maximum possible.", memberId);
+
+        if (currentAssignmentSize < maxAssignmentSize)
+            return true;
+
+        for (Uuid topicId : assignedTopicIds) {
+            if (canTopicParticipateInReassignment(topicId))
+                return true;
+        }
+        return false;
+    }
+
+    /**
+     * Determine if the current assignment is a balanced one.
+     *
+     * @return true if the given assignment is balanced; false otherwise.
+     */
+    private boolean isBalanced() {
```

Comment:
nit: In `performReassignments()` we iterate through partitions/members that we can reassign to others. Whereas in `isBalanced()` we iterate through members that can receive partitions. I feel it would be simpler if they follow the same pattern. It's not super important, but i got confused while reading the code

Reply:
makes sense, but it might also be better if we check balance differently from how we reassign the partitions, I can add it as a follow up if necessary